### PR TITLE
PI: Use iterative DFS in PdfWriter._sweep_indirect_references

### LINKED ISSUE #1036: Refactor writer._sweep_indirect_references: _idnum_hash and extern_map
In #207 we received valuable feedback that we might have code duplication: https://github.com/py-pdf/PyPDF2/pull/207#pullrequestreview-1021195338

It was not clarified within that PR as the PR improved PyPDF2 and I wanted to avoid a further delay (after 7 years of waiting!) with merging it.

This issue is only here as a reminder to have a look at it.

### PR DESCRIPTION
**Pull request by Hatell**
* Recursive Depth-first search (DFS) was changed to iterative DFS
* Removed `PdfWriter.external_reference_map` and calculate hash from every referred object and use that to detect duplicate objects.
* In several cases, the warning "Unable to resolve .*, returning NullObject instead" is no longer necessary.
* Bugfix: Recalculate all parents hashes when a dictionary or array object value changes

Closes #351
Closes #1036

### CODE REVIEW

**Comment by Hatell** [comment]
There is two tests with issues:
- test_sweep_recursion1
- test_sweep_recursion2

I did research them and those tests "succeeded" because this _sweep_indirect_references hit recursionlimit. And it happens because PDF has a linked list over 1000 items.

**Comment by MartinThoma** [comment]
Maybe we should work on getting https://github.com/py-pdf/PyPDF2/pull/351 ready first so that we don't hit the recursion limit anymore?

**Comment by Hatell** [comment]
This can be done in iterative algorithm like that.

**Comment by Hatell** [comment]
Now this is transformed to iterative version.

Some tests needed update because warnings was not raised any more.

**Comment by MartinThoma** [comment]
Wow, this is amazing @Hatell ! Thank you :pray: :hugs: 

I will review it today, but it might take to the evening :-)

**Comment by Hatell** [comment]
This is now ready for testing.

Main changes is:
- Recursive DFS was changed to this [DFS_iterative](https://en.wikipedia.org/wiki/Depth-first_search) algorithm.
- Removed extern_map and calculate hash from every referred object and use that to detect duplicate objects.

One fix need to be done to recalculate all parents hash if dictionary or array object value changes.

**Comment by Hatell** [comment]
I think I solved this issue for recalculating hashes when updating a dictionary or array object.

**Comment by MartinThoma** [comment]
Thank you so much for all the effort @Hatell !

I've adjusted the title of the PR and the first message of it. I will use them for the squash commit to represent all of the changes done here. Feel free to adjust if you think there should be something added / adjusted.

**Comment by MartinThoma** [comment]
If you want, you can also remove the

> # TODO: This test looks like an infinite loop.

in `test_merger.py`

File: PyPDF2/_writer.py [hunk 1]
```diff
@@ -871,70 +856,94 @@ def _sweep_indirect_references(
...
             NullObject,
         ],
-    ) -> Union[Any, StreamObject]:
-        if isinstance(data, DictionaryObject):
-            for key, value in list(data.items()):
-                value = self._sweep_indirect_references(extern_map, value)
-                if isinstance(value, StreamObject):
+    ) -> None:
+        stack: Any = collections.deque()
+        discovered = list()
```
Comment (MartinThoma) [inline PyPDF2/_writer.py:861]:
```suggestion
        discovered = []
```

File: PyPDF2/_writer.py [hunk 2]
```diff
@@ -871,70 +856,94 @@ def _sweep_indirect_references(
...
+        parent = None
+        grant_parents: List[PdfObject] = list()
```
Comment (MartinThoma) [inline PyPDF2/_writer.py:863]:
```suggestion
        grant_parents: List[PdfObject] = []
```

File: PyPDF2/_writer.py [hunk 3]
```diff
@@ -871,70 +856,94 @@ def _sweep_indirect_references(
...
-                            "returning NullObject instead",
-                            PdfReadWarning,
-                        )
-                        return NullObject()
-                return newobj
+                    data = self._resolve_indirect_object(self._add_object(data))
+
+                update_hashes = []
+
+                # Data changed what mead hash value changed
```
Comment (MartinThoma) [inline PyPDF2/_writer.py:892]:
```suggestion
                # Data changed and thus the hash value changed
```

File: PyPDF2/_writer.py [hunk 4]
```diff
@@ -871,70 +856,94 @@ def _sweep_indirect_references(
...
+                        if indirect_ref_obj is not None:
+                            self._idnum_hash[indirect_ref_obj.hash_value()] = indirect_ref
+            # while len(stack)
+
+    def _resolve_indirect_object(self, data: IndirectObject) -> IndirectObject:
+        """
+        Resolves indirect object to this pdf indirect objects.
+
+        If it is a new object then it is added to self._objects and new idnum is given and
+        generation is always 0.
```
Comment (MartinThoma) [inline PyPDF2/_writer.py:918]: This just prevents a line break in the GitHub diff view:

```suggestion
        If it is a new object then it is added to self._objects
        and new idnum is given and generation is always 0.
```

File: PyPDF2/_writer.py [hunk 5]
```diff
@@ -871,70 +856,94 @@ def _sweep_indirect_references(
             TextStringObject,
             NullObject,
         ],
-    ) -> Union[Any, StreamObject]:
-        if isinstance(data, DictionaryObject):
-            for key, value in list(data.items()):
-                value = self._sweep_indirect_references(extern_map, value)
-                if isinstance(value, StreamObject):
+    ) -> None:
+        stack: Any = collections.deque()
```
Comment (MartinThoma) [inline PyPDF2/_writer.py:860]:
```suggestion
        stack: Deque[
            Tuple[
                Any,
                Optional[Any],
                Any,
                List[PdfObject],
            ]
        ] = collections.deque()
```

File: PyPDF2/_writer.py [hunk 6]
```diff
@@ -28,6 +28,7 @@
 # POSSIBILITY OF SUCH DAMAGE.
 
 import codecs
+import collections
```
Comment (MartinThoma) [inline PyPDF2/_writer.py:31]:
```suggestion
import collections
from typing import Deque
```

File: PyPDF2/_writer.py (hunk 1 above)
Comment (MartinThoma) [inline PyPDF2/_writer.py:861]:
```suggestion
        discovered = []
```

File: PyPDF2/_writer.py [hunk 7]
```diff
@@ -871,70 +856,94 @@ def _sweep_indirect_references(
...
+                # Update old hash value to new hash value
+                for old_hash in update_hashes:
+                    indirect_ref = self._idnum_hash.pop(old_hash, None)
+
+                    if indirect_ref is not None:
+                        indirect_ref_obj = indirect_ref.get_object()
+
+                        if indirect_ref_obj is not None:
+                            self._idnum_hash[indirect_ref_obj.hash_value()] = indirect_ref
+            # while len(stack)
```
Comment (MartinThoma) [inline PyPDF2/_writer.py:911]:
```suggestion
```

**Comment by MartinThoma** [comment]
I'm currently letting a bigger text run through. So far, it looks good. I'm still a tiny bit worried as this is such a core part of PyPDF2 :sweat_smile:

**Comment by Hatell** [comment]
Great and thanks for help.

**Comment by MartinThoma** [comment]
Thank you for your contribution :heart: I'll make a release in a couple of hours

**Pull request merged**